core/src/main/java/org/infinispan/interceptors/impl/CacheLoaderInterceptor.java

```diff
@@ -347,45 +354,78 @@ protected boolean canLoad(Object key) {
 
-      CompletionStage<Void> stage = null;
-      if (entry != null) {
-         entryFactory.wrapExternalEntry(ctx, key, entry, true, cmd instanceof WriteCommand);
-
-         if (isLoadedValue != null && isLoadedValue) {
-            Object value = entry.getValue();
-            // FIXME: There's no point to trigger the entryLoaded/Activated event twice.
-            stage = sendNotification(key, value, true, ctx, cmd);
-            if (CompletionStages.isCompletedSuccessfully(stage)) {
-               stage = sendNotification(key, value, false, ctx, cmd);
-            } else {
-               stage = stage.thenCompose(v -> sendNotification(key, value, false, ctx, cmd));
+      CompletableFuture<InternalCacheEntry<K, V>> cf = new CompletableFuture<>();
+
+      CompletionStage<InternalCacheEntry<K, V>> otherCF = pendingLoads.putIfAbsent(key, cf);
+
+      // Another thread is currently retrieving the key from the store
+      if (otherCF == null) {
+         // Make sure we clean up our pendingLoads properly - note this is done before notifying any waiters
+         cf.whenComplete((ignore, t) -> pendingLoads.remove(key));
+         otherCF = cf;
+         int segment = SegmentSpecificCommand.extractSegment(cmd, key, partitioner);
+         CompletionStage<MarshallableEntry<K, V>> stage = PersistenceUtil.loadEntryAsync(persistenceManager, key,
+               segment, ctx, true);
+         stage.whenComplete((me, t) -> {
+            if (t != null) {
+               cf.completeExceptionally(t);
+               return;
             }
-         }
-      }
-      CacheEntry contextEntry = ctx.lookupEntry(key);
-      if (contextEntry instanceof MVCCEntry) {
-         ((MVCCEntry) contextEntry).setLoaded(true);
+            try {
+               if (me != null) {
+                  InternalCacheEntry<K, V> ice = PersistenceUtil.convert(me, iceFactory);
+                  if (getStatisticsEnabled()) {
+                     cacheLoads.incrementAndGet();
+                  }
+                  if (trace) {
+                     log.tracef("Loaded entry: %s for key %s from store and attempting to insert into data container",
+                           ice, key);
+                  }
+                  // TODO: remove could cause an issue as it removes from store which could have read before and then this
+                  // puts the value back in erroneously
+                  PersistenceUtil.putIfAbsentDataContainer(ice, dataContainer, timeService, segment);
+                  V value = ice.getValue();
+                  CompletionStage<Void> notificationStage = sendNotification(key, value, true, ctx, cmd);
+                  if (CompletionStages.isCompletedSuccessfully(notificationStage)) {
+                     notificationStage = sendNotification(key, value, false, ctx, cmd);
+                  } else {
+                     notificationStage = notificationStage.thenCompose(v -> sendNotification(key, value, false, ctx, cmd));
```

Comment:
The if/else avoids the allocation if there are no listeners, which is quite common. Since this is a persistence notification though I could simplify it to always do the allocation as the cost of the load should far outweigh the allocation.

Reply:
Minor, but we can go further here and remove the `notificationStage` var now we don't have the if; i.e.:

```java
sendNotification(key, value, true, ctx, cmd)
                        .thenCompose(v -> sendNotification(key, value, false, ctx, cmd))
                        .whenComplete((v, innerT) -> {
                           if (innerT != null) {
                              cf.completeExceptionally(innerT);
                           } else {
                              cf.complete(ice);
                           }
                        });
```